[3.x] i18n: Make property paths and categories translatable

timothyqiu opened the pull request:
`3.x` l10n is more complete than `master`, thus easier to test, so this PR targets `3.x`. I'll do a `master` version after this is discussed :)

I did some Ad-hoc localization for testing locally. When `interface/editor/translate_properties` is on, which is the default, the editor looks like this:

![ksnip_20220301-153147](https://user-images.githubusercontent.com/372476/156125220-96edca32-cbc0-4328-8460-52c543143539.png)

* As you can see from the screenshot, VRAM and GLES2 are correctly capitalized. This makes `msgid`s in the `POT` file easier to read. Capitalization logic is from (and thus supersedes) #32734.
    * Translation happens after capitalization.
    * Translation only happen when the new editor setting `interface/editor/translate_properties` is on.
* Tooltips are added to section headers.
    * When `translate_properties` is on, shows the original name.
    * When `translate_properties` is off, shows the translated name.
* In `extract.py`:
    * Replaced character by character line parsing with regular expression matching.
        * Extracting property paths are as easy as adding additional expressions.
        * I tested extraction without the property path patterns, the result is the same as the original approach. Except one string, which is fixed in #58627.
    * Capitalization data are also extracted from C++ code with regular expression.

timothyqiu committed:
i18n: Make property paths and sections translatable

Co-Authored-By: Hugo Locurcio <hugo.locurcio@hugo.pro>

timothyqiu commented on editor/translations/extract.py:
```diff
@@ -52,6 +62,36 @@
+
+def _process_editor_string(name):
+    # See String::capitalize().
+    # fmt: off
+    capitalized = " ".join(
+        part.title()
+        for part in capitalize_re.sub("_", name).replace("_", " ").split()
+    )
+    # fmt: on
```
I turned of `black` for this block. Because it wants me to put these on a single line.

Keeping these in multiple lines is much more readable I think.

timothyqiu commented:
Updated the regular expression for `GLOBAL_DEF` so that overrides like `audio/mix_rate.web` won't be matched.

bruvzg commented on editor/editor_sectioned_inspector.cpp:
```diff
@@ -266,7 +269,8 @@ void SectionedInspector::update_category_list() {
 				section_map[metasection] = ms;
-				ms->set_text(0, sectionarr[i].capitalize());
+				ms->set_text(0, EditorStringProcessor::get_singleton()->process_name(sectionarr[i]));
+				ms->set_tooltip(0, sectionarr[i]);
```
I would set a tooltip text to the translated name (or translated and raw names on the two lines) when `translate_properties` is `false`. This way, translation will be always available, even if you prefer raw names in the interface.

timothyqiu commented:
Updated how section tooltip works, it now looks like this:

| `translate_properties` on | `translate_properties` off |
| --------------------------------- | ---------------------------------|
|  ![ksnip_20220301-152903](https://user-images.githubusercontent.com/372476/156124731-06794504-d96a-4473-915b-1d6fb5ad7f46.png) | ![ksnip_20220301-152927](https://user-images.githubusercontent.com/372476/156124740-339f9d72-2b35-4b28-9a5a-35bd9377c1da.png) |

timothyqiu commented:
Added property name capitalization & translation to Editor Feature Profile dialog.

| `translate_properties` on | `translate_properties` off |
| --------------------------------- | ---------------------------------|
|  ![ksnip_20220302-160759](https://user-images.githubusercontent.com/372476/156320942-89144860-5d82-4662-bba8-67cf83a426a5.png) | ![ksnip_20220302-160827](https://user-images.githubusercontent.com/372476/156320952-f90ed683-ef8a-4f33-919d-00e09beaca22.png) |

akien-mga commented on editor/editor_string_processor.cpp:
```diff
@@ -0,0 +1,109 @@
+
+String EditorStringProcessor::process_name(const String &p_name) const {
+	const String capitalized_string = _capitalize_name(p_name);
+	if (EDITOR_GET("interface/editor/translate_properties")) {
```
Would it be worth caching and updating only on `NOTIFICATION_EDITOR_SETTINGS_CHANGED`?

timothyqiu replied:
I don't think it's necessary currently. The additonal settings check & localization are similar to two hash table lookups.

`process_name()` is also only used to initialize UI and re-fetched on `NOTIFICATION_EDITOR_SETTINGS_CHANGED`.

akien-mga commented on editor/editor_string_processor.cpp:
```diff
@@ -0,0 +1,109 @@
+}
+
+EditorStringProcessor::EditorStringProcessor() {
+	singleton = this;
```
Would it be worth adding an `ERR_FAIL_COND(singleton);` to make sure that we don't create two instances by mistake?

akien-mga commented on editor/editor_string_processor.cpp:
```diff
@@ -0,0 +1,109 @@
+EditorStringProcessor::EditorStringProcessor() {
+	singleton = this;
+
+	capitalize_string_remaps["2d"] = "2D";
```
You should add a comment that this is parsed in `editor/translations/extract.py` with a regex, so the map name and value definition format should be kept synced with the regex.

akien-mga commented on editor/editor_string_processor.cpp:
```diff
@@ -0,0 +1,109 @@
+EditorStringProcessor *EditorStringProcessor::singleton = nullptr;
+
+String EditorStringProcessor::_capitalize_name(const String &p_name) const {
+	String capitalized_string = p_name.capitalize();
+
+	// Fix the casing of a few strings commonly found in editor property/setting names
+	for (Map<String, String>::Element *E = capitalize_string_remaps.front(); E; E = E->next()) {
+		capitalized_string = capitalized_string.replace(E->key(), E->value());
+	}
```
It feels a bit hacky IMO to do replacements based on the wrongly capitalized output instead of handling the input directly (i.e. map the input `s3tc` and not the output `S 3 T C` to the correct form `S3TC`), but I don't see how to do it without making things less efficient and more cumbersome, so I guess it's fine as is :)

akien-mga commented on editor/editor_string_processor.h:
```diff
@@ -0,0 +1,58 @@
+
+#include "scene/main/node.h"
+
+class EditorStringProcessor : public Node {
```
I'm not convinced by the name, as it sounds like it might be something that would be used for all strings in the editor. Not sure what would be a better name, maybe `EditorPropertyNameProcessor`?

akien-mga commented on editor/editor_string_processor.cpp:
```diff
@@ -0,0 +1,109 @@
+String EditorStringProcessor::_capitalize_name(const String &p_name) const {
+	String capitalized_string = p_name.capitalize();
+
+	// Fix the casing of a few strings commonly found in editor property/setting names
```
Nitpick: dot at EOL.

akien-mga commented:
Looks pretty good to me overall! Added some review comments, mostly nitpicks. I do think the general approach is good :)

As a side note for the tooltips, it could be useful to have the full property paths (e.g. `application/config` for the "Config" section under "Application", or `application/config/name` for the property). But that's beyond the scope of this PR IMO, and maybe something to think about more for the general UX of these editors (e.g. to make sure that it's easy to copy a property name to clipboard, instead of just having it in a tooltip that would need to be typed manually).

timothyqiu commented on editor/editor_string_processor.h:
```diff
@@ -0,0 +1,58 @@
+class EditorStringProcessor : public Node {
+	GDCLASS(EditorStringProcessor, Node);
+
+	static EditorStringProcessor *singleton;
```
This is the declaration of the static member. It's defined & initialized to `nullptr` here:

https://github.com/godotengine/godot/blob/dfd778550b6336df759705608290616d0e53fad9/editor/editor_string_processor.cpp#L35

This approach is adapted by all static variables in the codebase currently.

timothyqiu commented:
Updated according to review and added some necessary remapping found during testing.

| From | To |
| ---- | -- |
| Gi Probe | GI Probe |
| Bvh | BVH |
| Cpu | CPU |
| Ik | IK |
| Png | PNG |
| Sdk | SDK |
| Vector 2 | Vector2 |

Checked the extracted messages, there are no false positives.

Calinou commented:
@timothyqiu You can find more English remappings here: https://github.com/godotengine/godot/pull/32734

timothyqiu commented:
@Calinou Yeah, those are what I used as the initial remappings. I saw that PR is several months old and some changes are required, so I did the changes here and listed you as the co-author :)

akien-mga commented:
Thanks!
